packages/fx-core/src/component/generator/templates/templateGenerator.ts

```diff
@@ -0,0 +1,223 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { hooks } from "@feathersjs/hooks/lib";
+import { Context, FxError, Inputs, Result, err, ok } from "@microsoft/teamsfx-api";
+import { TelemetryEvent, TelemetryProperty } from "../../../common/telemetry";
+import { convertToAlphanumericOnly } from "../../../common/utils";
+import { ProgressMessages, ProgressTitles } from "../../messages";
+import { ActionContext, ActionExecutionMW } from "../../middleware/actionExecutionMW";
+import { commonTemplateName, componentName } from "../constant";
+import { enableTestToolByDefault, isNewProjectTypeEnabled } from "../../../common/featureFlags";
+import {
+  CapabilityOptions,
+  MeArchitectureOptions,
+  ProgrammingLanguage,
+  QuestionNames,
+} from "../../../question";
+import { Generator, templateDefaultOnActionError } from "../generator";
+import { convertToLangKey, renderTemplateFileData, renderTemplateFileName } from "../utils";
+import { merge } from "lodash";
+import { GeneratorContext, TemplateActionSeq } from "../generatorAction";
+import { TemplateInfo } from "./templateInfo";
+import { Feature2TemplateName } from "./templateNames";
+
+export interface TemplateGenerator {
+  activate(ctx: Context, inputs: Inputs): boolean;
+  run(
+    ctx: Context,
+    inputs: Inputs,
+    destinationPath: string,
+    actionContext?: ActionContext
+  ): Promise<Result<undefined, FxError>>;
+  getTemplateInfos(
+    ctx: Context,
+    inputs: Inputs,
+    actionContext?: ActionContext
+  ): Promise<Result<TemplateInfo[], FxError>>;
+  post(
+    ctx: Context,
+    inputs: Inputs,
+    destinationPath: string,
+    actionContext?: ActionContext
+  ): Promise<Result<undefined, FxError>>;
+}
+
+export class DefaultTemplateGenerator implements TemplateGenerator {
+  componentName = componentName;
+
+  public activate(ctx: Context, inputs: Inputs): boolean {
+    return Object.keys(Feature2TemplateName).some((feature) =>
+      feature.startsWith(inputs.capabilities)
+    );
+  }
+
+  public getTemplateInfos(
+    ctx: Context,
+    inputs: Inputs,
+    actionContext?: ActionContext
+  ): Promise<Result<TemplateInfo[], FxError>> {
+    const templateName = this.getTemplateName(inputs);
+    const language = inputs[QuestionNames.ProgrammingLanguage] as ProgrammingLanguage;
+    const variables = this.getDefaultReplaceMap(inputs);
+    return Promise.resolve(ok([{ templateName, language, variables }]));
+  }
+
+  @hooks([
+    ActionExecutionMW({
+      enableProgressBar: true,
+      progressTitle: ProgressTitles.create,
+      progressSteps: 1,
+      enableTelemetry: true,
+      telemetryEventName: TelemetryEvent.GenerateTemplate,
+    }),
+  ])
+  public async run(
```

Comment:
It seems that run() is the coordinator of all functions of generator provider and there are only 3 extension points: activate(), mapTemplateInfos() and post(). Is there a better way to manage this so that developers will not override run() flow by accident?

Reply:
In TypeScript, there is no way to prevent a child class from overriding a public method of its parent class. I think this can only be done by documentation.